lantern_extras/src/bm25_agg.rs

```diff
@@ -0,0 +1,780 @@
+use crate::bloom::Bloom as PGBloom;
+use crate::{BM25_DEFAULT_APPROXIMATION_THRESHHOLD, BM25_DEFAULT_B, BM25_DEFAULT_K1};
+use fastbloom::BloomFilter;
+use std::collections::HashMap;
+
+use pgrx::datum::Internal;
+use pgrx::{pg_sys, prelude::*, AnyElement};
+
+#[derive(Copy, Clone, Default, Debug)]
+#[allow(non_camel_case_types)]
+pub struct bm25_agg_base;
+
+const HASHMAP_DEFAULT_CAPACITY: usize = 1000;
+// type used in doc_ids, fqs, lens array computations
+// all arrays are cast to this type before computations
+// todo:: q:: making this i64 does not seem to make things any slower.
+// I would expect it to significantly slow things down because of the extra copy/cast/write cycle
+type TargetInteger = i32;
+#[derive(Debug)]
+struct BM25InternalState {
+    data: Option<HashMap<TargetInteger, f32>>,
+    corpus_size: Option<u64>,
+    avg_doc_len: Option<f32>,
+    limit: Option<usize>,
+    // bm25 algorithm constants
+    k1: f32,
+    b: f32,
+    // approximate bm25 parameters
+    blooms: Vec<(f32, BloomFilter)>,
+}
+
+impl Default for BM25InternalState {
+    fn default() -> Self {
+        BM25InternalState {
+            data: Some(HashMap::with_capacity(HASHMAP_DEFAULT_CAPACITY)),
+            limit: None,
+            corpus_size: None,
+            avg_doc_len: None,
+            k1: BM25_DEFAULT_K1.get() as f32,
+            b: BM25_DEFAULT_B.get() as f32,
+            blooms: Vec::new(),
+        }
+    }
+}
+
+extension_sql!(
+    "\
+CREATE TYPE bm25result AS (
+    doc_id int8,
+    bm25 float4
+);",
+    name = "create_bm25result_type",
+    bootstrap
+);
+// we want to avoid repeating this in every overload before
+// but at the same time type Alias = ... does not seem to work with composite_type!
+const BM25RESULT_COMPOSITE_TYPE: &str = "bm25result";
+type BM25ResultSQLType = Vec<Option<pgrx::composite_type!('static, BM25RESULT_COMPOSITE_TYPE)>>;
+
+#[pg_extern(immutable, parallel_safe)]
+fn bm25_score(table_fqn: String, document: String, query: String) -> f32 {
+    let document_stemmed =
+        Spi::get_one::<Vec<String>>(&format!("SELECT text_to_stem_array('{}')", document))
+            .expect("Failed to get stemmed document")
+            .expect("Stemmed document was NULL");
+    let table_fqn = table_fqn.to_string();
+    let (corpus_size, avg_doc_len) = match Spi::get_two::<i32, f32>(&format!("SELECT term_freq AS corpus_size, (doc_ids_len / 100.0)::real AS avg_doc_len FROM {}_bm25 WHERE term IS NULL;", table_fqn))
+        .expect("Failed to get corpus size") {
+        (Some(corpus_size), Some(avg_doc_len)) => (corpus_size as u64, avg_doc_len),
+        _ => panic!("Failed to get corpus size and avg doc len"),
+    };
+    let term_data: Vec<(String, f32)> = Spi::connect(|client| {
+        client
+            .select(
+                &format!(
+                "SELECT term, term_freq FROM {}_bm25 WHERE term = ANY(text_to_stem_array('{}'));",
```

Comment:
do we need to `quote_ident` the table name to support upper-case table names or is it guaranteed to be lowercase?